pkg/pillar/hardware/smart.go

```diff
@@ -203,18 +221,33 @@ func GetInfoFromSCSIDisk(diskName string) (*types.DiskSmartInfo, error) {
 }
 
 // GetSerialNumberForDisk tеakes a disk name (from dev directory,
-// for example /dev/sda) as input and return serial number
+// for example /dev/sda or /dev/sda1) as input and return serial number
 func GetSerialNumberForDisk(diskName string) (string, error) {
-	block, err := ghw.Block()
+	dev, err := smart.Open(diskName)
 	if err != nil {
-		return "", fmt.Errorf("error getting block storage info: %v", err)
+		return "", fmt.Errorf("disk with name: %s have unknown type", diskName)
 	}
+	diskType := dev.Type()
+	dev.Close()
 
-	for _, disk := range block.Disks {
-		if fmt.Sprintf("/dev/%v", disk.Name) == diskName {
-			return disk.SerialNumber, nil
+	var diskSmartInfo *types.DiskSmartInfo
+
+	if diskType == "sata" {
+		diskSmartInfo, err = GetInfoFromSATAdisk(diskName)
+		if err != nil {
+			return "", err
+		}
+	} else if diskType == "nvme" {
+		diskSmartInfo, err = GetInfoFromNVMeDisk(diskName)
+		if err != nil {
+			return "", err
+		}
+	} else if diskType == "scsi" {
+		diskSmartInfo, err = GetInfoFromSCSIDisk(diskName)
+		if err != nil {
+			return "", err
 		}
 	}
```

Comment:
Last else branch (presumably with an error and returning "") is missing.
Alternatively call the unknown function (which doesn't set serial) and use diskSmartInfo.SerialNumber

Reply:
Corrected. I have added an error return